### PR TITLE
#10: Feat - LFU cache implementation with approximate counting.

### LINKED ISSUE #10: Add support for LFU eviction using Approximate Counting

### PR DESCRIPTION
**Pull request by Rits1272**
Fixes #10
- LFU cache eviction policy support in Dice

How LFU cache is implemented?
- The implementation is similar to Redis - [details](https://redis.io/docs/latest/develop/reference/eviction/#:~:text=is%20very%20simple.-,The%20new%20LFU%20mode,-Starting%20with%20Redis)
- The `access` counter of a key is piggybacked to the `LastAccessedAt` field in Dice Obj defined [here](https://github.com/DiceDB/dice/blob/d3e49e25cba3fad01225cb38fdde122049f4ec5e/internal/store/object.go#L10). The first 8 bits is allocated to maintaining the access counter and the last 24 bits are allocated to maintaining the last access time of a key
- The `access` counter is an approximate counter which is incremented logarithmically
- This PR does not include support for the `decay` mechanism as of yet.
- This LFU implementation do not add any additional memory footprint 

Issue: https://github.com/DiceDB/dice/issues/10

### CODE REVIEW

**Comment by Rits1272** [comment]
@JyotinderSingh please have a look. TIA!

File: internal/store/evictionpool.go [hunk 1]
```diff
@@ -30,6 +33,26 @@ func (a ByIdleTime) Less(i, j int) bool {
 	return GetIdleTime(a[i].lastAccessedAt) > GetIdleTime(a[j].lastAccessedAt)
 }
 
+func (a ByCounterAndIdleTime) Len() int {
+	return len(a)
+}
+
+func (a ByCounterAndIdleTime) Swap(i, j int) {
```
Comment (JyotinderSingh) [inline internal/store/evictionpool.go:40]: Where is this used?

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for introducing a new eviction algorithm to Dice @Rits1272! Please add comprehensive tests for these changes.

**Comment by arpitbbhayani** [comment]
@Rits1272 Please squash your commits before you push. You can do this once all changes are done.

File: internal/store/evictionpool.go (hunk 1 above)
Comment (kushal0511-not) [reply internal/store/evictionpool.go:40]: I think it is for implementing Sort interface for **ByCounterAndIdleTime** and using it like this 
```
sort.Sort(ByCounterAndIdleTime(pq.pool))
```
below at line 68

**Comment by JyotinderSingh** [comment]
Hey @Rits1272, could you please address the comments and fix any conflicts? This PR has been open a while, let's try to get it merged.

**Comment by Rits1272** [comment]
@JyotinderSingh didn't get time this week, but will work on adding test cases, resolving conflicts and comments by this weekend

**Comment by Rits1272** [comment]
@JyotinderSingh sorry for the delay. I have added test cases, please review
Since the PR has gotten stale, there are many conflicts from previous commits. Can we use Squash and merge for this PR during merge to master?

**Review by JyotinderSingh — APPROVED** [review]
Thanks for adding support for LFU eviction algorithm, @Rits1272! The changes look good.
Approved

**Pull request merged**